[documentation] add more jvm and os guidelines

himanshug opened the pull request:
Fixes #4792

Reviewers: Feel free to suggest edits based on your experience.

drcrallen commented:
here are a few logging ones that are mentioned elsewhere in the docs I think:

```
-Djava.util.logging.manager=org.apache.logging.log4j.jul.LogManager
-Dorg.jboss.logging.provider=slf4j
-Dnet.spy.log.LoggerImpl=net.spy.memcached.compat.log.SLF4JLogger
-Dlog4j.shutdownCallbackRegistry=io.druid.common.config.Log4jShutdown
-Dlog4j.shutdownHookEnabled=true
```

The downside of this one is that middle managers have noisy error messages when running hadoop or spark stuff due to classloader miss-matches caused by the JUL --> log4j2 binding. Benign but noisy.

gianm commented on docs/content/operations/recommendations.md:
```diff
@@ -5,6 +5,39 @@ layout: doc_page
 Recommendations
 ===============
 
+# Some Genearl guidelines
```
general (spelling)

himanshug replied:
fixed

gianm commented on docs/content/operations/recommendations.md:
```diff
@@ -5,6 +5,39 @@ layout: doc_page
 
+# Some Genearl guidelines
+
+JVM Flags..
```
Two dots is kind of strange. Suggest a colon or making this a subheading.

himanshug replied:
using  a colon

gianm commented on docs/content/operations/recommendations.md:
```diff
@@ -5,6 +5,39 @@ layout: doc_page
+
+`ExitOnOutOfMemoryError` flag is only supported starting JDK 8u92 . For older versions, `-XX:OnOutOfMemoryError='kill -9 %p'` can be used.
+
+`MaxDirectMemorySize` restricts jvm from allocating more than specified limit, by setting it to unlimited jvm restriction is lifted and OS level memory limits would take effect.
```
Maybe add:

> It's still important to make sure that Druid is not configured to allocate more off-heap memory than your machine has available. Important settings here include `druid.processing.numThreads`, `druid.processing.numMergeBuffers`, and `druid.processing.buffer.sizeBytes`.

Just so people don't think lifting MaxDirectMemorySize to the sky absolves them of need to configure memory use.

himanshug replied:
added

gianm commented on docs/content/operations/recommendations.md:
```diff
@@ -5,6 +5,39 @@ layout: doc_page
+
+`MaxDirectMemorySize` restricts jvm from allocating more than specified limit, by setting it to unlimited jvm restriction is lifted and OS level memory limits would take effect.
+
+Please note that above flags are general guidelines only, Be cautious and feel free to change them if necessary for the specific deployment.
```
"only, Be" isn't grammatical. Suggest "only. Be"

himanshug replied:
fixed

gianm commented on docs/content/operations/recommendations.md:
```diff
@@ -5,6 +5,39 @@ layout: doc_page
+
+Please note that above flags are general guidelines only, Be cautious and feel free to change them if necessary for the specific deployment.
+
+Additionally, for large jvm heaps, here are few Garbage Collection efficiency guidelines that have been known to help in some cases.
```
"are a few" (grammar)

himanshug replied:
fixed

gianm commented on docs/content/operations/recommendations.md:
```diff
@@ -5,6 +5,39 @@ layout: doc_page
+
+Additionally, for large jvm heaps, here are few Garbage Collection efficiency guidelines that have been known to help in some cases.
+- Mount /tmp on tmpfs ( See http://www.evanjones.ca/jvm-mmap-pause.html )
+- On Disc-IO intensive nodes (e.g. Historical and MiddleManager), GC and Druid logs should be written to a different disk than where data is written.
```
"Disk" (spelling)

himanshug replied:
fixed

gianm commented on docs/content/operations/recommendations.md:
```diff
@@ -5,6 +5,39 @@ layout: doc_page
+-XX:+ExitOnOutOfMemoryError
+-XX:+HeapDumpOnOutOfMemoryError
+-XX:HeapDumpPath=/var/logs/druid/historical.hprof
+-XX:MaxDirectMemorySize=10240g
```
I usually recommend people set `-Djava.io.tmpdir` explicitly since sometimes /tmp is not where they actually want to store tmp files.

himanshug replied:
we have this set also, wasn't sure if its general enough... added with some explanation

himanshug commented:
@drcrallen @gianm added logging options and addressed comments. there are two more things that we have tried in the past in various clusters , not enough benefit analysis has been done to know if they have measurable impact. that is why I haven't added those in docs, but here they are for the sake of completeness

- try disabling jvm  biased locking in case of GC issues.
- use JBOD configuration instead of RAID on historicals to avoid RAID overhead specially if you don't have RAID controller hardware but a software based RAID is configured.

gianm commented:
I would say go ahead and add the JBOD thing. I've tested it (on EC2 at least) vs software RAID and it does improve performance.

himanshug commented:
@gianm i added both actually.. i think they are just options that user should be aware of and do what is necessary.

jon-wei commented:
going to go ahead and merge this for 0.11.0 since it's a doc patch and it's already approved
